src/MMALSharp.Processing/Processors/Motion/FrameDiffDriver.cs

```diff
@@ -0,0 +1,311 @@
+        /// <param name="config">The motion configuration object.</param>
+        /// <param name="onDetect">A callback when changes are detected.</param>
+        public FrameDiffDriver(MotionConfig config, Action onDetect)
+        {
+            _motionConfig = config;
+
+            _onDetect = onDetect;
+            this.OnDetectEnabled = _onDetect != null;
+
+            _testFrameAge = new Stopwatch();
+            this.LastDetectionEvent = new Stopwatch();
+        }
+
+        // These are used for temporary local performance-testing; uncomment these and four lines in
+        // the Apply method below, and see the Dispose method at the end of FrameBufferCaptureHandler.
+        // private Stopwatch frameTimer = new Stopwatch();
+        // internal long frameCounter;
+        // internal long totalElapsed;
+
+        /// <inheritdoc />
+        public override void Apply(ImageContext context)
+        {
+            _imageContext = context;
+
+            base.Apply(context);
+
+            if (context.Eos)
+            {
+                // if zero bytes buffered, EOS is the end of a physical input video filestream
+                if (this.WorkingData.Count > 0)
+                {
+                    if (!_fullTestFrame)
+                    {
+                        MMALLog.Logger.LogDebug("EOS reached for test frame.");
+
+                        _fullTestFrame = true;
+                        this.PrepareTestFrame();
+                    }
+                    else
+                    {
+                        MMALLog.Logger.LogDebug("Have full frame, checking for changes.");
+
+                        this.CurrentFrame = this.WorkingData.ToArray();
+
+                        // frameCounter++;
+                        // frameTimer.Restart();
+
+                        var detected = _motionConfig.MotionAlgorithm.DetectMotion(this, _frameMetrics);
+
+                        // frameTimer.Stop();
+                        // totalElapsed += frameTimer.ElapsedMilliseconds;
+
+                        if (detected)
+                        {
+                            this.LastDetectionEvent.Restart();
+
+                            if(this.OnDetectEnabled)
+                            {
+                                _onDetect?.Invoke();
+                            }
+                        }
+
+                        this.TryUpdateTestFrame();
+                    }
+                }
+                else
+                {
+                    MMALLog.Logger.LogDebug("EOS reached, no working data buffered");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Resets the state of the buffers so that a new test frame is
+        /// stored. Also resets any state in the motion detection algorithm.
+        /// </summary>
+        public void ResetAnalyser()
+        {
+            this.FullFrame = false;
+            _fullTestFrame = false;
+            this.WorkingData = new List<byte>();
```

Comment:
Could we see any further benefit to moving `this.WorkingData` into a `byte[]` instead of a `List<byte>`?

Reply:
No, internally `List<byte>` is just an array, so we'd just end up doing the same work (resizing etc). One thing that might help, though, is to think about initializing the list to a certain size. I believe the default is just 10 entries and there is some sort of scale-up algorithm, it may be helpful to set it to a large size up front, given the amount of data we're working with. I'll open an issue so we remember to consider it later on. I think there are multiple places in the code where that idea might apply.